src/modm/platform/uart/sam/uart_hal_impl.hpp.in

```diff
@@ -0,0 +1,170 @@
+}
+
+void
+modm::platform::{{ name }}::reset()
+{
+	{{ peripheral }}->USART.CTRLA.bit.SWRST = true;	// Uart Reset
+	while({{ peripheral }}->USART.SYNCBUSY.bit.SWRST); // Wait for sync.
+}
+
+template<class SystemClock, modm::baudrate_t baudrate>
+void modm_always_inline
+modm::platform::{{ name }}::initialize(Parity parity)
+{
+	// Enable peripheral clock in power manager.
+	PM->APBCMASK.bit.{{ peripheral }}_ = true;
+	// Init clock using GCLK0 as source.
+	GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID_{{ peripheral }}_CORE |
+                      GCLK_CLKCTRL_GEN_GCLK0 |
+                      GCLK_CLKCTRL_CLKEN ;
+	while (GCLK->STATUS.bit.SYNCBUSY);
+	// Reset USART
+	reset();
+	// Set clock mode internal
+	{{ peripheral }}->USART.CTRLA.bit.MODE = 0x1;
+	// Set asynchronous mode
+	{{ peripheral }}->USART.CTRLA.bit.CMODE = 0x0;
+	// Set rx/tx pins
+	{{ peripheral }}->USART.CTRLA.bit.RXPO = 0x3; // Pad3
+	{{ peripheral }}->USART.CTRLA.bit.TXPO = 0x1; // Pad2;
+	// Set character size to 8 bits
+	{{ peripheral }}->USART.CTRLB.bit.CHSIZE = 0x0;
+	// Set Data order to LSB first
+	{{ peripheral }}->USART.CTRLA.bit.DORD = 0x1;
+	// Set parity
+	setParity(parity);
+	// Set stop bits to 1
+	{{ peripheral }}->USART.CTRLB.bit.SBMODE = 0x0;
+	// fractional baud rate
+	{{ peripheral }}->USART.CTRLA.bit.SAMPR = 0x1;
+	// set baudrate (multiply by 8 to calc fractional part)
+	uint16_t baudTimes8 = (SystemClock::Frequency * 8) / (16 * baudrate);
+	SERCOM0->USART.BAUD.FRAC.BAUD = baudTimes8 / 8;
+	SERCOM0->USART.BAUD.FRAC.FP = baudTimes8 % 8;
+	// uint16_t baud = SystemClock::Frequency / (16 * baudrate);
+	// SERCOM0->USART.BAUD.reg = baud;
+	setReceiverEnable(true);
+	setTransmitterEnable(true);
+	// Enable USART
+	enable();
+}
+
+
+void
+modm::platform::{{ name }}::write(uint8_t data)
+{
+	{{ peripheral }}->USART.DATA.reg = data;
+}
+
+void
+modm::platform::{{ name }}::read(uint8_t &data)
+{
+	data = {{ peripheral }}->USART.DATA.reg;
+}
+
+void
+modm::platform::{{ name }}::setTransmitterEnable(const bool enable)
+{
+	{{ peripheral }}->USART.CTRLB.bit.TXEN = enable;
+}
+
+void
+modm::platform::{{ name }}::setReceiverEnable(bool enable)
+{
+	{{ peripheral }}->USART.CTRLB.bit.RXEN = enable;
+}
+
+void
+modm::platform::{{ name }}::enableOperation()
+{
+	enable();
+}
+
+void
+modm::platform::{{ name }}::disableOperation()
+{
+	disable();
+}
+
+bool
+modm::platform::{{ name }}::isReceiveRegisterNotEmpty()
+{
+	return {{ peripheral }}->USART.INTFLAG.bit.RXC;
+}
+
+bool
+modm::platform::{{ name }}::isTransmitRegisterEmpty()
+{
+	return {{ peripheral }}->USART.INTFLAG.bit.DRE;
+}
+
+template<template<modm::platform::Peripheral _> class RxPin>
+void modm_always_inline
+modm::platform::{{ name }}::setRxPin()
+{
+	using modm::platform::Peripheral;
+	{{ peripheral }}->USART.CTRLA.bit.RXPO =
+	    uint32_t(RxPin<Peripheral::{{ sercom }}>::RxPad);
+	PORT->Group[0].PINCFG[11].bit.PMUXEN = true;
+  PORT->Group[0].PMUX[11 >> 1].reg |= PORT_PMUX_PMUXE_C;
```

Comment:
This looks hardcoded? You can use the `RxPin::Gpio::port` and `RxPin::Gpio::pin` cast to int here.
I would think the `PMUX` would be implemented in the `RxPin::connect()` method using the data from modm-devices (if it's complete that is).